File: tests/test-streaming-frontend.cpp
#include "moonshine-streaming.h"
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>

// Generate a sine wave test signal
static std::vector<float> make_sine_wave(float freq_hz, float duration_s, int sample_rate = 16000) {
    int n = (int)(duration_s * sample_rate);
    std::vector<float> audio(n);
    for (int i = 0; i < n; i++) {
        audio[i] = 0.5f * sinf(2.0f * M_PI * freq_hz * i / sample_rate);
    }
    return audio;
}

int main(int argc, char ** argv) {
    const char * model_path = nullptr;
    bool verbose = false;
    const char * dump_path = nullptr;

    for (int i = 1; i < argc; i++) {
        if ((strcmp(argv[i], "-m") == 0 || strcmp(argv[i], "--model") == 0) && i + 1 < argc) {
            model_path = argv[++i];
        } else if (strcmp(argv[i], "-v") == 0 || strcmp(argv[i], "--verbose") == 0) {
            verbose = true;
        } else if (strcmp(argv[i], "--dump") == 0 && i + 1 < argc) {
            dump_path = argv[++i];
        }
    }

    if (!model_path) {
        fprintf(stderr, "usage: %s -m <model.gguf> [-v] [--dump output.bin]\n", argv[0]);
        return 1;
    }

    // Load model
    printf("Loading model: %s\n", model_path);
    struct moonshine_streaming_context * ctx = moonshine_streaming_init(model_path);
    if (!ctx) {
        fprintf(stderr, "Failed to load model\n");
        return 1;
    }

    if (verbose) {
        moonshine_streaming_print_model_info(ctx);
    }

    // Generate 1 second of 440Hz sine wave
    auto audio = make_sine_wave(440.0f, 1.0f);
    printf("Audio: %zu samples (%.2f sec)\n", audio.size(), audio.size() / 16000.0f);

    // Run frontend
    float * features = nullptr;
    int seq_len = 0;
    int hidden_dim = 0;

    int ret = moonshine_streaming_frontend(ctx, audio.data(), (int)audio.size(),
                                            &features, &seq_len, &hidden_dim);
    if (ret != 0) {
        fprintf(stderr, "Frontend failed\n");
        moonshine_streaming_free(ctx);
        return 1;
    }

    printf("Frontend output: [%d, %d] (hidden x seq_len)\n", hidden_dim, seq_len);

    // Print some statistics
    float min_val = features[0], max_val = features[0], sum = 0;
    for (int i = 0; i < hidden_dim * seq_len; i++) {
        if (features[i] < min_val) min_val = features[i];
        if (features[i] > max_val) max_val = features[i];
        sum += features[i];
    }
    float mean = sum / (hidden_dim * seq_len);
    printf("Stats: min=%.4f max=%.4f mean=%.4f\n", min_val, max_val, mean);

    // Print first few values for comparison with Python
    printf("First 10 values: ");
    for (int i = 0; i < 10 && i < hidden_dim * seq_len; i++) {
        printf("%.6f ", features[i]);
    }
    printf("\n");

    // Dump to binary file for comparison
    if (dump_path) {
        FILE * f = fopen(dump_path, "wb");
        if (f) {
            int32_t dims[2] = { hidden_dim, seq_len };
            fwrite(dims, sizeof(int32_t), 2, f);
            fwrite(features, sizeof(float), hidden_dim * seq_len, f);
            fclose(f);
            printf("Dumped features to %s\n", dump_path);
        }
    }

    free(features);
    moonshine_streaming_free(ctx);
    printf("OK\n");
    return 0;
}

File: models/test_streaming_frontend.py
#!/usr/bin/env python3
"""
Compare C++ streaming frontend output against Python reference.

Usage:
    # Step 1: Run C++ and dump features
    ./build/bin/test-streaming-frontend -m models/moonshine-streaming-tiny-f32.gguf --dump /tmp/cpp_frontend.bin

    # Step 2: Run this script to compare
    python models/test_streaming_frontend.py --cpp-dump /tmp/cpp_frontend.bin

    # Or just run Python reference without comparison:
    python models/test_streaming_frontend.py
"""

import argparse
import struct
import sys

import numpy as np
import torch
from transformers import AutoModel


def make_sine_wave(freq_hz=440.0, duration_s=1.0, sample_rate=16000):
    """Generate same test signal as C++ test."""
    n = int(duration_s * sample_rate)
    t = np.arange(n, dtype=np.float32) / sample_rate
    return 0.5 * np.sin(2.0 * np.pi * freq_hz * t).astype(np.float32)


def run_python_frontend(audio: np.ndarray) -> np.ndarray:
    """Run the Python reference frontend (encoder embedder only)."""
    model = AutoModel.from_pretrained(
        "usefulsensors/moonshine-streaming-tiny",
        trust_remote_code=True,
    )
    model.eval()

    embedder = model.encoder.embedder

    with torch.no_grad():
        audio_tensor = torch.from_numpy(audio).unsqueeze(0)  # [1, N]
        features, _ = embedder(audio_tensor)
        # features: [1, seq_len, hidden] in PyTorch
        return features.squeeze(0).numpy()  # [seq_len, hidden]


def load_cpp_dump(path: str):
    """Load binary dump from C++ test: [hidden_dim(i32), seq_len(i32), float32 data...]"""
    with open(path, "rb") as f:
        hidden_dim, seq_len = struct.unpack("ii", f.read(8))
        data = np.frombuffer(f.read(), dtype=np.float32)
        # ggml stores ne[0]=hidden, ne[1]=seq in column-major order
        # which is equivalent to [seq_len, hidden] in numpy row-major
        return data.reshape(seq_len, hidden_dim)  # → [seq_len, hidden]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--cpp-dump", help="Path to C++ binary dump for comparison")
    args = parser.parse_args()

    audio = make_sine_wave(440.0, 1.0)
    print(f"Audio: {len(audio)} samples ({len(audio)/16000:.2f} sec)")

    # Run Python reference
    print("Running Python frontend...")
    py_features = run_python_frontend(audio)
    print(f"Python output: {py_features.shape} (seq_len x hidden)")
    print(f"Stats: min={py_features.min():.4f} max={py_features.max():.4f} mean={py_features.mean():.4f}")
    print(f"First 10 values (row-major): {py_features.ravel()[:10]}")

    if args.cpp_dump:
        print(f"\nLoading C++ dump: {args.cpp_dump}")
        cpp_features = load_cpp_dump(args.cpp_dump)
        print(f"C++ output: {cpp_features.shape}")

        if py_features.shape != cpp_features.shape:
            print(f"SHAPE MISMATCH: Python {py_features.shape} vs C++ {cpp_features.shape}")
            sys.exit(1)

        diff = np.abs(py_features - cpp_features)
        print(f"\nComparison:")
        print(f"  Max abs diff:  {diff.max():.6e}")
        print(f"  Mean abs diff: {diff.mean():.6e}")
        print(f"  RMS diff:      {np.sqrt((diff**2).mean()):.6e}")

        # Print first few values side by side
        print(f"\nFirst 5 values (Python vs C++):")
        for i in range(min(5, py_features.shape[0])):
            for j in range(min(5, py_features.shape[1])):
                print(f"  [{i},{j}] py={py_features[i,j]:+.6f}  cpp={cpp_features[i,j]:+.6f}  diff={diff[i,j]:.2e}")

        if diff.max() < 1e-4:
            print("\nPASS: outputs match within tolerance (1e-4)")
        elif diff.max() < 1e-3:
            print("\nWARN: outputs close but max diff > 1e-4")
        else:
            print("\nFAIL: outputs differ significantly")
            sys.exit(1)
    else:
        # Just dump Python output for manual comparison
        print(f"\nPython first 10 (col-major, matching C++ layout):")
        # C++ stores [hidden, seq_len], so column-major traversal = iterate hidden first
        py_colmajor = py_features.T  # [hidden, seq_len]
        print(f"  {py_colmajor.ravel()[:10]}")


if __name__ == "__main__":
    main()
